packages/gatsby/src/utils/page-data.js

```diff
@@ -35,7 +37,57 @@ const write = async ({ publicDir }, page, result) => {
   await fs.outputFile(filePath, bodyStr)
 }
 
+const getChangedPageDataKeys = (store, cacheData) =>
+  new Promise(resolve => {
+    if (cacheData.webpackCompilationHash !== store.webpackCompilationHash) {
+      resolve([...store.pages.keys()])
+      return
+    }
+    if (cacheData.pageData && store.pageData) {
+      const pageKeys = []
+      store.pageData.forEach((value, key) => {
+        if (!cacheData.pageData.has(key)) {
+          pageKeys.push(key)
+        } else {
+          const newPageData = JSON.stringify(value)
+          const previousPageData = JSON.stringify(cacheData.pageData.get(key))
+
+          if (newPageData !== previousPageData) {
+            pageKeys.push(key)
+          }
+        }
+      })
+      resolve(pageKeys)
+      return
+    }
+
+    resolve([...store.pages.keys()])
+  })
+
+const removePreviousPageData = (directory, store, cacheData) =>
+  new Promise(resolve => {
```

Comment:
Also doesn't need to be a promise if you do fs.removeSync anyway.

Also probably better to separate deleting pages from this function - this can return a list of files to delete, a function in build can then actually remove them all at once.

Reply:
Done 👍 I've moved the deleting of the pages to the build step.